lib/ansible/module_utils/vmware.py

```diff
@@ -828,41 +831,174 @@ def is_vcenter(self):
+                if self.params['folder'] is None:
+                    self.module.fail_json(msg="Multiple virtual machines with same name [%s] found, "
+                                              "Folder value is a required parameter to find uniqueness "
+                                              "of the virtual machine" % self.params['name'],
+                                          details="Please see documentation of the vmware_guest module "
+                                                  "for folder parameter.")
+
+                # Get folder path where virtual machine is located
+                # User provided folder where user thinks virtual machine is present
+                user_folder = self.params['folder']
+                # User defined datacenter
+                user_defined_dc = self.params['datacenter']
+                # User defined datacenter's object
+                datacenter_obj = find_datacenter_by_name(self.content, self.params['datacenter'])
+                # Get Path for Datacenter
+                dcpath = compile_folder_path_for_object(vobj=datacenter_obj)
+
+                # Nested folder does not return trailing /
+                if not dcpath.endswith('/'):
+                    dcpath += '/'
+
+                if user_folder in [None, '', '/']:
+                    # User provided blank value or
+                    # User provided only root value, we fail
+                    self.module.fail_json(msg="vmware_guest found multiple virtual machines with same "
+                                              "name [%s], please specify folder path other than blank "
+                                              "or '/'" % self.params['name'])
+                elif user_folder.startswith('/vm/'):
+                    # User provided nested folder under VMware default vm folder i.e. folder = /vm/india/finance
+                    user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)
+                else:
+                    # User defined datacenter is not nested i.e. dcpath = '/' , or
+                    # User defined datacenter is nested i.e. dcpath = '/F0/DC0' or
+                    # User provided folder starts with / and datacenter i.e. folder = /ha-datacenter/ or
+                    # User defined folder starts with datacenter without '/' i.e.
+                    # folder = DC0/vm/india/finance or
+                    # folder = DC0/vm
+                    user_desired_path = user_folder
+
+                for vm in vms:
+                    # Check if user has provided same path as virtual machine
+                    actual_vm_folder_path = self.get_vm_path(content=self.content, vm_name=vm)
+                    if not actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc)):
+                        continue
+                    if user_desired_path in actual_vm_folder_path:
```

Comment:
@Akasurde 
```
+                if user_folder in [None, '', '/']:
+                    # User provided blank value or
+                    # User provided only root value, we fail
+                    self.module.fail_json(msg="vmware_guest found multiple virtual machines with same "
+                                              "name [%s], please specify folder path other than blank "
+                                              "or '/'" % self.params['name'])
+                elif user_folder == '/vm' or user_folder.startswith('/vm/'):
+                    # User provided VMware default vm folder with / or
+                    # User provided nested folder under VMware default vm folder i.e. folder = /vm/india/finance
+                    user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)
+                elif user_folder == 'vm' or user_folder == '%s' % user_defined_dc:
+                    # User provided VMware default vm folder
+                    # User defined folder is datacenter i.e. folder = 'DC0'
+                    user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)
+                else:
+                    # User defined datacenter is not nested i.e. dcpath = '/' , or
+                    # User defined datacenter is nested i.e. dcpath = '/F0/DC0' or
+                    # User provided folder starts with / and datacenter i.e. folder = /ha-datacenter/ or
+                    # User defined folder starts with datacenter without '/' i.e.
+                    # folder = DC0/vm/india/finance or
+                    # folder = DC0/vm
+                    user_desired_path = user_folder
+
+                for vm in vms:
+                    # Check if user has provided same path as virtual machine
+                    actual_vm_folder_path = self.get_vm_path(content=self.content, vm_name=vm)
+                    if not actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc)):
+                        continue
+                    if user_desired_path in actual_vm_folder_path:
+                        vm_obj = vm
+                        break
```


 * `if user_folder in [None, '', '/']` → fail
OK, I am good with that.

* `elif user_folder == '/vm' …` → `user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)`

If user_folder == '/vm', user_desired_path will be = to "/dcpath/user_defined_dc/vm". So if  `actual_vm_folder_path.startswith("%s%s" % (dcpath, user_defined_dc))`, user_desired_path will always be in actual_vm_folder_path.

The case `user_folder == '/vm'` is useless since we get the same result with user_desired_path = user_folder.

* `or user_folder.startswith('/vm/'):` → `user_desired_path = "%s%s%s" % (dcpath, user_defined_dc, user_folder)`

It bring a bit of precision (compared to user_desired_path = user_folder), so Ok for me

* `elif user_folder == 'vm' …` → `user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)`

Same as before, the case `user_folder == 'vm'` is useless since we get the same result with user_desired_path = user_folder.

* `or user_folder == '%s' % user_defined_dc:` → `user_desired_path = "%s%s/vm" % (dcpath, user_defined_dc)`

Side note: `user_folder == '%s' % user_defined_dc` should be written `user_folder == user_defined_dc`.

Same as before, the case `user_folder == user_defined_dc` is useless since we get the same result with user_desired_path = user_folder.

Reply:
@tchernomax I updated patch as per your suggestion. Could you please take a look at new patch ?